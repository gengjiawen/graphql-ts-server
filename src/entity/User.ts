import { BaseEntity, BeforeInsert, Column, Entity, PrimaryGeneratedColumn } from 'typeorm'
import * as bcrypt from 'bcryptjs'
import { ObjectType, Field, ID } from 'type-graphql'

@ObjectType()
@Entity('users')
export class User extends BaseEntity {
  @Field(() => ID)
  @PrimaryGeneratedColumn('uuid')
  id: string

  @Field()
  @Column('varchar', { length: 255 })
  email: string

  @Field()
  @Column('text')
  password: string

  @Field()
  @Column('boolean', { default: false })
  confirmed: boolean

  @BeforeInsert()
  async hashPassword() {
    this.password = await bcrypt.hash(this.password, 10)
  }
}
