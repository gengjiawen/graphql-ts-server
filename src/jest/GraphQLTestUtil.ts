import axios, { AxiosResponse } from 'axios'

export function graphQLRequest(payload: string) {
  const url = process.env.TEST_HOST as string
  return axios.post(url, { query: payload })
}

export function mapErrorsMessage(response: AxiosResponse) {
  const errors: any[] = response.data.errors
  return errors.map(i => i.message)
}
