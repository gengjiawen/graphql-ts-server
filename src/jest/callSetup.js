require('ts-node/register')
module.exports = require('./setup').default
