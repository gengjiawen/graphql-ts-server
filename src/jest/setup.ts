import { startServer } from '../StartServer'
import { AddressInfo } from 'net'

export default async function() {
  if (!process.env.TEST_HOST) {
    const app = await startServer()
    const { port } = app.address() as AddressInfo
    const testHost = `http://localhost:${port}`
    console.log(testHost)
    process.env.TEST_HOST = testHost
  }
}
