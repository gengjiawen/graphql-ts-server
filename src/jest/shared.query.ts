export const registerMutation = (e: string, p: string) => `
mutation {
  register(email: "${e}", password: "${p}") {
    token
  }
}`

export const loginMutation = (e: string, p: string) => `
mutation {
  login(email: "${e}", password: "${p}") {
    token
  }
}`
