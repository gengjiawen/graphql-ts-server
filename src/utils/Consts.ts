export const JWT_SECRET = 'winds of winter'
