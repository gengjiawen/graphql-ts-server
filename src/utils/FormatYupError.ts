import { ValidationError } from 'yup'

export const formatYupError = (error: ValidationError) => {
  const errors: Array<{ path: string; message: string }> = []
  error.inner.forEach(e => {
    errors.push({
      path: e.path,
      message: e.message,
    })
  })

  return errors
}
