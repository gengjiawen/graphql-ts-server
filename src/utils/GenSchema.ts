import { GraphQLSchema } from 'graphql'
import * as fs from 'fs'
import * as path from 'path'
import { importSchema } from 'graphql-import'
import { makeExecutableSchema, mergeSchemas } from 'graphql-tools'

export function genSchema() {
  const schemas: GraphQLSchema[] = []
  const modulePath = path.join(__dirname, '../modules')
  console.log(modulePath)
  const folders = fs.readdirSync(modulePath)
  folders.forEach(folder => {
    const schemaPath = `${modulePath}/${folder}/schema.graphql`
    if (!fs.existsSync(schemaPath)) {
      console.log(`schema path: ${schemaPath} has no graphql definition, skipping`)
      return
    }
    const { resolvers } = require(`${modulePath}/${folder}/resolvers`)
    const typeDefs = importSchema(schemaPath)
    schemas.push(makeExecutableSchema({ resolvers, typeDefs }))
  })
  const schema = mergeSchemas({ schemas })
  return schema
}
