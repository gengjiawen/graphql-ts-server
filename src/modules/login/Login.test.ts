import { graphQLRequest, mapErrorsMessage } from '../../jest/GraphQLTestUtil'
import { User } from '../../entity/User'
import { loginMutation, registerMutation } from '../../jest/shared.query'
import { createTypeormConn } from '../../utils/CreateTyepeormConn'

const email = 'jack_login@test.com'
const password = '123456'

beforeAll(async () => {
  await createTypeormConn()
})

test('login successfully', async () => {
  const registerRes: any = await graphQLRequest(registerMutation(email, password))
  expect(registerRes.data.data.register.token).not.toBeNull()
  await User.update({ email }, { confirmed: true })

  // login success
  const loginRes: any = await graphQLRequest(loginMutation(email, password))
  expect(loginRes.data.data.login.token).not.toBeNull()
  expect(loginRes.data.data.errors).toBeUndefined()

  // login with wrong password
  const loginWrongPassRes: any = await graphQLRequest(loginMutation(email, '123'))
  expect(loginWrongPassRes.data.data.login).toBeNull()
  expect(mapErrorsMessage(loginWrongPassRes)).toMatchSnapshot()
})
