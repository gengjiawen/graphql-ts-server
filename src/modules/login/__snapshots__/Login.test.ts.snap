// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login with wrong password 1`] = `
Array [
  "invalid login",
]
`;
